### PR TITLE
[CMDBUF] Fix incorrect handling of shared local mem args in CUDA/HIP

### PR DESCRIPTION
**Pull request by Bensuo**
- Fix handling of local mem args in CUDA/HIP for command buffer update
- Add conformance tests which check updating local memory args and work size

### CODE REVIEW

**Comment by Bensuo** [comment]
DPC++ PR: https://github.com/intel/llvm/pull/15920

File: test/conformance/exp_command_buffer/update/local_memory_update.cpp [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+    // New Y at index 4
+    new_input_descs[1] = {
+        UR_STRUCTURE_TYPE_EXP_COMMAND_BUFFER_UPDATE_POINTER_ARG_DESC, // stype
+        nullptr,                                                      // pNext
+        4,               // argIndex
+        nullptr,         // pProperties
+        &shared_ptrs[4], // pArgValue
+    };
+
+    size_t new_local_size = local_size * 2;
```
Comment (EwanC) [inline test/conformance/exp_command_buffer/update/local_memory_update.cpp:250]: Could define this before line 216 and use it there too

File: test/conformance/exp_command_buffer/update/local_memory_update.cpp
Comment (EwanC) [inline test/conformance/exp_command_buffer/update/local_memory_update.cpp:1]: Think we'll have to add the new tests to the `exp_command_buffer_adapter_native_cpu.match` file, all the

File: test/conformance/exp_command_buffer/update/local_memory_update.cpp (hunk 1 above)
Comment (Bensuo) [reply test/conformance/exp_command_buffer/update/local_memory_update.cpp:250]: As an interesting consesquence of making this change it occurs to me that having the `argSize` member of the value update be `uint32_t` is probably a mistake, and it should be `size_t` to match the argSize passed when actually setting arguments.

Currently with the change now it requires a cast. I'll see if I can change it without breaking everything.

Comment (Bensuo) [reply test/conformance/exp_command_buffer/update/local_memory_update.cpp]: It doesn't seem to break anything so I've just gone ahead and made that change.

**Comment by EwanC** [comment]
Code change look fine to me, but there's some HIP command-buffer CTS tests failing in CI which might not be spurious

**Review by EwanC — APPROVED** [review]

**Comment by EwanC** [comment]
@oneapi-src/unified-runtime-maintain This is ready for review, there is a CUDA E2E CI fail but that's not reflected in the CI for the [matching DPC++ PR](https://github.com/intel/llvm/pull/15920) which is green and also seems to be in other open PRs like https://github.com/oneapi-src/unified-runtime/pull/2263

```
Failed Tests (14):
  SYCL :: Basic/kernel_info.cpp
  SYCL :: Basic/kernel_info_attr.cpp
  SYCL :: Basic/launch_queries/max_num_work_groups.cpp
  SYCL :: DeviceCodeSplit/split-per-kernel.cpp
  SYCL :: DeviceCodeSplit/split-per-source-main.cpp
  SYCL :: Graph/Explicit/spec_constants_kernel_bundle_api.cpp
  SYCL :: Graph/RecordReplay/spec_constants_kernel_bundle_api.cpp
  SYCL :: GroupAlgorithm/back_to_back_collectives.cpp
  SYCL :: GroupAlgorithm/root_group.cpp
  SYCL :: KernelAndProgram/kernel-bundle-find-run.cpp
  SYCL :: SpecConstants/2020/kernel-bundle-api.cpp
  SYCL :: SubGroup/info.cpp
  SYCL :: USM/P2P/p2p_access.cpp
  SYCL :: syclcompat/util/max_active_work_groups_per_cu.cpp
```

**Review by callumfare — APPROVED** [review]

**Pull request merged**